flocker/control/_diffing.py

```diff
@@ -0,0 +1,247 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+# -*- test-case-name: flocker.control.test.test_diffing -*-
+
+from pyrsistent import (
+    PClass,
+    PMap,
+    PVector,
+    PSet,
+    field,
+    freeze,
+    pvector,
+)
+
+from zope.interface import Interface, implementer
+
+
+class _DiffChange(Interface):
+    """
+    Interface for a diff change.
+
+    This is simply something that can be applied to an object to create a new
+    object.
+
+    This interface is created more to simply documentation than for any of the
+    actual zope.interface mechanisms.
+    """
+
+    def apply(obj):
+        """
+        Apply this diff change to the passed in object and return a new object
+        that is obj with the ``self`` diff applied.
+
+        :param object obj: The object to apply the diff to.
+
+        :returns: A new object that is the passed in object with the diff
+            applied.
+        """
+        pass
+
+
+@implementer(_DiffChange)
+class _Remove(PClass):
+    """
+    A ``_DiffChange`` that removes an object from a ``PSet`` or a key from a
+    ``PMap`` inside a nested object tree.
+
+    :ivar path: The path in the nested object tree of the object to be removed
+        from the import set.
+    """
+
+    path = field(
+        type=PVector,
+        factory=freeze
+    )
+
+    def apply(self, obj):
+        obj_path = self.path[:-1]
+        removal_path = self.path[-1]
+        return obj.transform(obj_path, lambda o: o.remove(removal_path))
+
+
+@implementer(_DiffChange)
+class _Set(PClass):
+    """
+    A ``_DiffChange`` that sets a field in a ``PClass`` or sets a key in a
+    ``PMap``.
+
+    :ivar path: The path in the nested object to the field/key to be set to a
+        new value.
+
+    :ivar val: The value to set the field/key to.
+    """
+    path = field(
+        type=PVector,
+        factory=freeze
+    )
+    val = field()
+
+    def apply(self, obj):
+        return obj.transform(self.path, self.val)
+
+
+@implementer(_DiffChange)
+class _Add(PClass):
+    """
+    A ``_DiffChange`` that adds an item to a ``PSet``.
+
+    :ivar path: The path to the set to which the item will be added.
+
+    :ivar item: The item to be added to the set.
+    """
+    path = field(
+        type=PVector,
+        factory=freeze
+    )
+    item = field()
+
+    def apply(self, obj):
+        return obj.transform(self.path, lambda x: x.add(self.item))
+
+
+@implementer(_DiffChange)
+class _Diff(PClass):
+    """
+    A ``_DiffChange`` that is simply the serial application of other diff
+    changes.
+
+    This is the object that external modules get and use to apply diffs to
+    objects.
+
+    :ivar changes: A vector of ``_DiffChange`` s that represent a diff between
+        two objects.
+    """
+
+    changes = field(
+        type=PVector,
+        factory=freeze
+    )
+
+    def apply(self, obj):
+        for c in self.changes:
+            obj = c.apply(obj)
+        return obj
```

Comment:
:question: This looks like a situation where it'd be faster to operate on ``obj.evolver()`` rather than the immutable object.